barretenberg/cpp/src/barretenberg/dsl/acir_format/ec_operations.cpp

```diff
@@ -13,19 +14,126 @@
+                                           builder);
+            } else {
+
+                cycle_group_ct input1_point;
+                cycle_group_ct input2_point;
+                // all or nothing: the inputs must be all constant or all witness. Cf #1108 for more details.
+                if (!input.input1_x.is_constant || !input.input1_y.is_constant || !input.input1_infinite.is_constant ||
+                    !input.input2_x.is_constant || !input.input2_y.is_constant || !input.input2_infinite.is_constant) {
+                    // One of the input is a witness, so we ensure that all inputs are witness, by creating witness for
+                    // constant values.
+                    input1_point = to_witness_grumpkin_point(input.input1_x,
+                                                             input.input1_y,
+                                                             input.input1_infinite,
+                                                             has_valid_witness_assignments,
+                                                             true,
+                                                             builder);
+                    input2_point = to_witness_grumpkin_point(input.input2_x,
+                                                             input.input2_y,
+                                                             input.input2_infinite,
+                                                             has_valid_witness_assignments,
+                                                             false,
+                                                             builder);
+
+                } else {
+                    input1_point = to_grumpkin_point(input.input1_x,
+                                                     input.input1_y,
+                                                     input.input1_infinite,
+                                                     has_valid_witness_assignments,
+                                                     true,
+                                                     builder);
+                    input2_point = to_grumpkin_point(input.input2_x,
+                                                     input.input2_y,
+                                                     input.input2_infinite,
+                                                     has_valid_witness_assignments,
+                                                     false,
+                                                     builder);
+                }
+                // both points are not infinity, so we can use unconditional_add
+                if (has_valid_witness_assignments) {
+                    // Runtime check that the inputs have not the same x coordinate, as assumed by the function.
+                    ASSERT(input1_point.x.get_value() != input2_point.x.get_value());
+                }
+                result = input1_point.unconditional_add(input2_point);
```

Comment:
Why are you using unconditional add? Are you checking that the inputs are different? If the values are the same, this could allow you to create any point as a result

Reply:
This is the whole point of the PR, the checks have been moved on the Noir side so that we can avoid them in some cases. See this comment in the PR description:
"It will allow Aztec protocol circuit to use directly the opcode when the 'safety' is already known (and implied by previous operations)."

Reply:
You shouldn't remove the difference!=zero check. It is always necessary. There is no case that it is not necessary, when you are performing non-doubling additions.

Reply:
I disagree, else the function `unconditional_add()` would not exist. See the comments of the function:
 *        Only use this method if you know the x-coordinates of the operands cannot collide and none of the operands is a point at infinity

For instance it is used for batch mul:
 *         If `unconditional_add = true`, we use `::unconditional_add` instead of `::checked_unconditional_add`. Use with caution! Only should be `true` if we're doing an ULTRA fixed-base MSM so we know the points cannot collide with the offset generators.

Reply:
I will add a setting then.

Reply:
Done